### ISSUE TITLE
Style is not being updated when a prop is changed

### ISSUE THREAD

**Issue by brunolemos** [styled-components/styled-components#304]
## Version
1.1.3

## Behavior on v1.1.1 (no bugs)
A styled component is using props and the theme for styling. When any of them is changed, the style should be updated. (_optimization: only if the prop is being used for styling_)

[![styled-components-before-bug-v1-1-1.gif](https://s28.postimg.org/akwalp6wt/styled_components_before_bug.gif)](https://postimg.org/image/bzxvaf7zt/)

## Behavior on v1.1.2 (bug on theme change)
The style is updated when a prop changes, but **not** when the `theme` change. See:

> Just for comparison. Already reported and fixed on #261

[![styled-components-theme-bug.gif](https://s30.postimg.org/3zam9sp0x/styled_components_theme_bug.gif)](https://postimg.org/image/hsyyyuhm5/)

## Behavior on v1.1.3 (bug on prop change)
The style is **not** being updated when a prop changes, only when the `theme` change. See:

> Reproduction on v1.1.3:

[![styled-components-prop-bug-v-1-1-3.gif](https://s27.postimg.org/k1f0mhssj/styled_components_prop_bug.gif)](https://postimg.org/image/vqj0agjr3/)

## Relevant code:
```js
const Star = styled.View`
  width: 50;
  height: 50;
  background-color: ${({ starred }) => starred ? '#edb800' : 'gray'};
`;
```

```js
<Star starred={starred} />
```

## Full reproduction code
```js
import React, { Component } from 'react';
import styled, { ThemeProvider } from 'styled-components/native';
import { AppRegistry, TouchableOpacity } from 'react-native';

const StyledPage = styled.View`
  flex: 1;
  align-items: center;
  justify-content: center;
  background-color: ${({ theme }) => theme.backgroundColor};
`;

const StyledText = styled.Text`
  color: ${({ theme }) => theme.textColor};
`;

const Star = styled.View`
  width: 50;
  height: 50;
  background-color: ${({ starred }) => starred ? '#edb800' : 'gray'};
`;

const darkTheme = { backgroundColor: '#333333', textColor: '#ffffff' };
const lightTheme = { backgroundColor: '#eeeeee', textColor: '#111111' };

export default class App extends Component {
  state = {
    theme: darkTheme,
    starred: false,
  };

  render() {
    const { starred, theme } = this.state;
    const nextTheme = theme === darkTheme ? lightTheme : darkTheme;

    return (
      <ThemeProvider theme={theme}>
        <StyledPage>
          <Star starred={starred} />
          
          <TouchableOpacity onPress={() => this.setState({ starred: !starred })}>
            <StyledText>
              {starred ? 'Starred!' : 'Not starred'}
            </StyledText>
          </TouchableOpacity>

          <TouchableOpacity onPress={() => this.setState({ theme: nextTheme })}>
            <StyledText>
              Current theme: {theme === lightTheme ? 'Light' : 'Dark'}
            </StyledText>
          </TouchableOpacity>
        </StyledPage>
      </ThemeProvider>
    );
  }
}

AppRegistry.registerComponent('styledbug', () => App);
```

**Comment by k15a** [comment]
Could you check if this is just React native and if this is a new bug with 1.1.3?

**Comment by brunolemos** [comment]
@k15a yes, it appears to be only in react native and only after v1.1.3.
I updated my first comment with more gifs for the last 3 versions, check it out.

Here is the behaviour in the web:
[![styled-components-bug-web.gif](https://s27.postimg.org/p93c5llmr/styled_components_bug_web.gif)](https://postimg.org/image/5r8opnoov/)

_It's rendering the previous theme, not the current one, but that's another issue, not this issue._

**Comment by k15a** [comment]
> It's rendering the previous theme, not the current one, but that's another issue.

Can't reproduce this: http://www.webpackbin.com/NkZCyQ47f

It's probably one of my PR's which caused this issue: https://github.com/styled-components/styled-components/pull/295 and https://github.com/styled-components/styled-components/pull/264
Because I changed `StyledComponent.js` and `StyledNativeComponent.js` in the same way I can't imagine what's causing this bug.

Unfortunately I've never used React native so I can't test this myself.

**Comment by brunolemos** [comment]
@k15a if you have a mac or can download an android emulator, all the code needed is in the first comment.

> Can't reproduce this: http://www.webpackbin.com/NkZCyQ47f

Sorry, the web version is fine on 1.1.3. I did not see my web project was using 1.1.1.

**Comment by mxstbr** [comment]
> I might try to fix it if no one else does first...

Please do @brunolemos, super sorry for the troubles and thanks for the detailled reproduction!

**Comment by simonas88** [comment]
I'm having the same issue as OP with v2.1.2 of styled-components and react-native v0.47.0

**Comment by philpl** [comment]
@simonas88 please open a new issue with the information that is provided in the template :)

This issue is closed and was fixed :) if a similar issue manifests itself on a different version it's definitely a different issue altogether

Also, commenting on an old issue notifies all previous participants which is not necessary ;)
